Index: mwlwifi-2018-12-18-c2c8244d/hif/fwcmd.c
===================================================================
--- mwlwifi-2018-12-18-c2c8244d.orig/hif/fwcmd.c
+++ mwlwifi-2018-12-18-c2c8244d/hif/fwcmd.c
@@ -2189,6 +2189,10 @@ int mwl_fwcmd_set_new_stn_add(struct iee
 		return -EIO;
 	}
 
+    pcmd->cmd_hdr.cmd = cpu_to_le16(HOSTCMD_CMD_SET_NEW_STN);
+    pcmd->cmd_hdr.len = cpu_to_le16(sizeof(*pcmd));
+    pcmd->cmd_hdr.macid = mwl_vif->macid;
+
 	if (vif->type == NL80211_IFTYPE_STATION) {
 		ether_addr_copy(pcmd->mac_addr, mwl_vif->sta_mac);
 		pcmd->aid = cpu_to_le16(sta->aid + 1);
@@ -2391,6 +2395,9 @@ int mwl_fwcmd_set_new_stn_del(struct iee
 	if (vif->type == NL80211_IFTYPE_STATION) {
 		ether_addr_copy(pcmd->mac_addr, mwl_vif->sta_mac);
 
+        pcmd->cmd_hdr.cmd = cpu_to_le16(HOSTCMD_CMD_SET_NEW_STN);
+        pcmd->cmd_hdr.len = cpu_to_le16(sizeof(*pcmd));
+
 		if (mwl_hif_exec_cmd(hw, HOSTCMD_CMD_SET_NEW_STN)) {
 			mutex_unlock(&priv->fwcmd_mutex);
 			return -EIO;
@@ -2541,6 +2548,10 @@ int mwl_fwcmd_update_encryption_enable(s
 		else
 			ether_addr_copy(pcmd->mac_addr, mwl_vif->bssid);
 
+        pcmd->cmd_hdr.cmd = cpu_to_le16(HOSTCMD_CMD_UPDATE_ENCRYPTION);
+        pcmd->cmd_hdr.len = cpu_to_le16(sizeof(*pcmd));
+        pcmd->cmd_hdr.macid = mwl_vif->macid;
+
 		if (mwl_hif_exec_cmd(hw, HOSTCMD_CMD_UPDATE_ENCRYPTION)) {
 			mutex_unlock(&priv->fwcmd_mutex);
 			return -EIO;
@@ -2635,6 +2646,10 @@ int mwl_fwcmd_encryption_set_key(struct
 			ether_addr_copy(pcmd->key_param.mac_addr,
 					mwl_vif->bssid);
 
+        pcmd->cmd_hdr.cmd = cpu_to_le16(HOSTCMD_CMD_UPDATE_ENCRYPTION);
+        pcmd->cmd_hdr.len = cpu_to_le16(sizeof(*pcmd));
+        pcmd->cmd_hdr.macid = mwl_vif->macid;
+
 		if (mwl_hif_exec_cmd(hw, HOSTCMD_CMD_UPDATE_ENCRYPTION)) {
 			mutex_unlock(&priv->fwcmd_mutex);
 			return -EIO;
@@ -3042,6 +3057,9 @@ int mwl_fwcmd_set_wsc_ie(struct ieee8021
 	}
 
 	pcmd->ie_type = cpu_to_le16(WSC_IE_SET_PROBE_RESPONSE);
+    pcmd->cmd_hdr.cmd = cpu_to_le16(HOSTCMD_CMD_SET_WSC_IE);
+    pcmd->cmd_hdr.len = cpu_to_le16(sizeof(*pcmd));
+    pcmd->len = cpu_to_le16(len);
 
 	if (mwl_hif_exec_cmd(hw, HOSTCMD_CMD_SET_WSC_IE)) {
 		mutex_unlock(&priv->fwcmd_mutex);
