Index: mwlwifi-2018-12-18-c2c8244d/hif/pcie/pcie.c
===================================================================
--- mwlwifi-2018-12-18-c2c8244d.orig/hif/pcie/pcie.c
+++ mwlwifi-2018-12-18-c2c8244d/hif/pcie/pcie.c
@@ -1566,6 +1566,7 @@ static int pcie_probe(struct pci_dev *pd
 	pci_set_drvdata(pdev, hw);
 
 	priv = hw->priv;
+        priv->heartbeat = 1;
 	priv->antenna_tx = pcie_chip_tbl[priv->chip_type].antenna_tx;
 	priv->antenna_rx = pcie_chip_tbl[priv->chip_type].antenna_rx;
 	pcie_priv = priv->hif.priv;
